streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java

```diff
@@ -64,7 +64,7 @@ public void setIfUnset(final SerdeGetter getter) {
 
         @Override
         public byte[] serialize(final String topic, final SubscriptionResponseWrapper<V> data) {
-            //{1-bit-isHashNull}{7-bits-version}{Optional-16-byte-Hash}{n-bytes serialized data}
+            //{1-bit-isHashNull}{7-bits-version}{4-bytes-primaryPartition}{Optional-16-byte-Hash}{n-bytes serialized data}
```

Comment:
I think we cannot just write the new format blindly. The receiving task may be executed on a different instance that might not have been ungraded yet during a rolling upgrade. -- We can only start to use the new format after all instances are upgraded, otherwise older instances would crash (or the upgrade can not be done in rolling bounce fashion what seen not acceptable).

Not sure right now how we can handle this case though: The coarse grained solution would be to use a config and let users do two rolling bounces (requires a KIP). Or we try to piggy-back on the rebalance metadata version, but I am not sure how easy it would be to do, given that we setup the serdes at a totally different layer and get access to the rebalance metadata version only during runtime.

Reply:
Yeah, this is indeed a problem. It's the same issue I documented here: https://issues.apache.org/jira/browse/KAFKA-10336

I was also thinking that the rebalance metadata version could give us a mechanism to coordinate this kind of thing across a rolling bounce. I documented the approach in that ticket, but never got a chance to work on it.

If we don't want to bite off a general solution to it, two rolling bounces might be the best compromise, as annoying as they are.

Alternatively, since we actually do check the version and throw an exception if we encounter it, it is "safe" (in that there's no data corruption) to just go ahead and let the exceptions happen during a rolling bounce. Not the nicest thing to do, though.

Reply:
What will happen when an old processor receives a v1 record and throws an exception? Will it throw that record on the floor or just block the processing until the upgrade?

As John suggested in [KAFKA-10336](https://issues.apache.org/jira/browse/KAFKA-10336), we can pass `UPGRADE_FROM` into the serde so that it produces v0 records if the option is set. I will also add `3.3` as a valid choice for that option. WDYT?